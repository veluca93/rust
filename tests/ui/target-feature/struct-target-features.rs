//@ only-x86_64
#![feature(struct_target_features)]
//~^ WARNING the feature `struct_target_features` is incomplete and may not be safe to use and/or cause compiler crashes
#![feature(target_feature_11)]

use std::arch::x86_64::*;

#[target_feature(enable = "avx")]
struct Invalid(u32);

#[target_feature(enable = "avx")]
struct Avx {}

#[target_feature(enable = "sse")]
struct Sse();

#[target_feature(enable = "avx")]
fn avx() {}

trait TFAssociatedType {
    type Assoc;
}

impl TFAssociatedType for () {
    type Assoc = Avx;
}

#[target_feature(from_args)]
fn avx_self(_: <() as TFAssociatedType>::Assoc) {
    avx();
}

#[target_feature(from_args)]
fn avx_avx(_: Avx) {
    avx();
}

#[target_feature(from_args)]
extern "C" fn bad_fun(_: Avx) {}

#[inline(always)]
//~^ ERROR cannot use `#[inline(always)]` with `#[target_feature]`
#[target_feature(from_args)]
fn inline_fun(_: Avx) {}
//~^ ERROR cannot use a struct with target features in a #[inline(always)] function

trait Simd {
    fn do_something(&self);
}

impl Simd for Avx {
    #[target_feature(from_args)]
    fn do_something(&self) {
        unsafe {
            println!("{:?}", _mm256_setzero_ps());
        }
    }
}

impl Simd for Sse {
    #[target_feature(from_args)]
    fn do_something(&self) {
        unsafe {
            println!("{:?}", _mm_setzero_ps());
        }
    }
}

#[target_feature(from_args)]
fn avxfn(_: &Avx) {
    // This is not unsafe because we already have the feature at function-level.
    let _ = Avx {};
}

fn main() {
    Avx {};
    //~^ ERROR initializing type `Avx` with `#[target_feature]` is unsafe and requires unsafe function or block [E0133]

    if is_x86_feature_detected!("avx") {
        let avx = unsafe { Avx {} };
        avxfn(&avx);
        avx.do_something();
    }
    if is_x86_feature_detected!("sse") {
        unsafe { Sse {} }.do_something();
    }
}
