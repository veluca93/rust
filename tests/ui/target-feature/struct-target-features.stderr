warning: the feature `struct_target_features` is incomplete and may not be safe to use and/or cause compiler crashes
  --> $DIR/struct-target-features.rs:2:12
   |
LL | #![feature(struct_target_features)]
   |            ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #129107 <https://github.com/rust-lang/rust/issues/129107> for more information
   = note: `#[warn(incomplete_features)]` on by default

error: cannot use a struct with target features in a #[inline(always)] function
  --> $DIR/struct-target-features.rs:44:1
   |
LL | fn inline_fun(_: Avx) {}
   | ^^^^^^^^^^^^^^^^^^^^^

error: cannot use `#[inline(always)]` with `#[target_feature]`
  --> $DIR/struct-target-features.rs:41:1
   |
LL | #[inline(always)]
   | ^^^^^^^^^^^^^^^^^

error[E0133]: initializing type `Avx` with `#[target_feature]` is unsafe and requires unsafe function or block
  --> $DIR/struct-target-features.rs:76:5
   |
LL |     Avx {};
   |     ^^^^^^ call to function with `#[target_feature]`
   |
   = note: the  target features being enabled in the build configuration does not remove the requirement to list them in `#[target_feature]`

error: aborting due to 3 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0133`.
